### PR TITLE
add stable var checking and gcflag to moc.js

### PR DESCRIPTION
**Pull request by chenyan-dfinity**
Tested in https://github.com/dfinity/motoko-playground/pull/91

### CODE REVIEW

**Comment by chenyan-dfinity** [comment]
@crusso Is there an easy way to get the stable signature from type checking or desugar? In the playground, it would be nice if we can get the signature before producing the Wasm.

**Comment by crusso** [comment]
There is the --stable-type types option, but that writes to <foo>.most. Or you can read it from the custom section.

File: src/js/common.ml [hunk 1]
```diff
@@ -57,21 +57,35 @@ let js_candid source =
...
+  js_result (Pipeline.compile_files mode true [source])
+    (fun (idl_prog, m) ->
+      let open CustomModule in
+      let sig_ = match m.motoko.stable_types with
+        | Some (_, txt) -> Js.some (Js.string txt)
+        | _ -> Js.null in
+      let candid = Idllib.Arrange_idl.string_of_prog idl_prog in
       let (_, wasm) = CustomModuleEncode.encode m in
       let constructor = Js.Unsafe.global##._Uint8Array in
       let code = constructor##from
```
Comment (crusso) [inline src/js/common.ml:80]: I know this old code, but is the code returned as a stream, not just a JS bytearray? Doesn't that make reading each byte really slow?

File: src/js/common.ml [hunk 2]
```diff
@@ -109,3 +123,10 @@ let set_actor_aliases entries =
   let aliases = Flags.actor_aliases in
   aliases := Flags.M.of_seq (Array.to_seq entries)  
 
+let gc_flags option =
+  match Js.to_string option with
+  | "force" -> Flags.force_gc := true
+  | "scheduling" -> Flags.force_gc := false
+  | "copying" -> Flags.gc_strategy := Mo_config.Flags.Copying
+  | "marking" -> Flags.gc_strategy := Mo_config.Flags.MarkCompact
```
Comment (crusso) [inline src/js/common.ml:131]: Would it make sense to use the same names as moc compiler options, just for sanity?

File: src/js/common.ml [hunk 3]
```diff
@@ -57,21 +57,35 @@ let js_candid source =
...
     | "wasi" -> Flags.WASIMode
     | "ic" -> Flags.ICMode
     | _ -> raise (Invalid_argument "js_compile_with: Unexpected mode")
   in
-  js_result (Pipeline.compile_files mode true [Js.to_string source])
-    (fun (_, m) ->
+  js_result (Pipeline.compile_files mode true [source])
+    (fun (idl_prog, m) ->
+      let open CustomModule in
+      let sig_ = match m.motoko.stable_types with
```
Comment (crusso) [inline src/js/common.ml:74]: You can also get the arg-less service type this way, should you need it.

File: src/js/common.ml (hunk 1 above)
Comment (chenyan-dfinity) [reply src/js/common.ml:80]: It's returning a bytearray/Uint8Array. The constructor is slow, but it's kind of the only way to produce a bytearray to JS. Other "faster" method breaks the content if it's not utf8 compatible.

File: src/js/common.ml (hunk 2 above)
Comment (chenyan-dfinity) [reply src/js/common.ml:131]: I think this is close enough. There is no sanity in JS...I just spent the last two days figuring out that `await setX(42)` in React doesn't guarantee x is 42 after the await...

**Pull request merged**